compiler/src/dotty/tools/dotc/plugins/Plugins.scala

```diff
@@ -0,0 +1,268 @@
+
+
+  /** A description of all the plugins that are loaded */
+  def pluginDescriptions: String =
+    roughPluginsList map (x => "%s - %s".format(x.name, x.description)) mkString "\n"
+
+  /** Summary of the options for all loaded plugins */
+  def pluginOptionsHelp: String =
+    (for (plug <- roughPluginsList ; help <- plug.optionsHelp) yield {
+      "\nOptions for plugin '%s':\n%s\n".format(plug.name, help)
+    }).mkString
+
+  /** Add plugin phases to phase plan */
+  def addPluginPhases(plan: List[List[Phase]])(implicit ctx: Context): List[List[Phase]] = {
+    // plugin-specific options.
+    // The user writes `-P:plugname:opt1,opt2`, but the plugin sees `List(opt1, opt2)`.
+    def options(plugin: Plugin): List[String] = {
+      def namec = plugin.name + ":"
+      ctx.settings.pluginOptions.value filter (_ startsWith namec) map (_ stripPrefix namec)
+    }
+
+    // schedule plugins according to ordering constraints
+    val pluginPhases = plugins.filter(!_.research).flatMap(plug => plug.init(options(plug)))
+    val updatedPlan = Plugins.schedule(plan, pluginPhases)
+
+    // add research plugins
+    plugins.filter(_.research).foldRight(updatedPlan) { (plug, plan) => plug.init(options(plug), plan) }
+  }
+}
+
+object Plugins {
+  /** Insert plugin phases in the right place of the phase plan
+   *
+   *  The scheduling makes sure the ordering constraints of plugin phases are satisfied.
+   *  If the ordering constraints are unsatisfiable, an exception is thrown.
+   *
+   *  Note: this algorithm is factored out for unit test.
+   */
+  def schedule(plan: List[List[Phase]], pluginPhases: List[PluginPhase]): List[List[Phase]] = {
+    import scala.collection.mutable.{ Map => MMap }
+    type OrderingReq = (Set[Class[_]], Set[Class[_]])
+
+    val orderRequirements = MMap[Class[_], OrderingReq]()
+    val primitivePhases   = plan.flatMap(ps => ps.map(_.getClass.asInstanceOf[Class[_]])).toSet
+
+    def isPrimitive(phase: Class[_]): Boolean = primitivePhases.contains(phase)
+
+    def constraintConflict(phase: Phase): String = {
+      val (runsAfter, runsBefore) = orderRequirements(phase.getClass)
+      s"""
+         |Ordering conflict for phase ${phase.phaseName}
+         |after: ${runsAfter.mkString("[", ", ", "]")}
+         |before: ${runsBefore.mkString("[", ", ", "]")}
+       """.stripMargin
+    }
+
+    // init ordering map, no propagation
+    pluginPhases.foreach { phase =>
+      val runsAfter  : Set[Class[_]] = phase.runsAfter.asInstanceOf[Set[Class[_]]]
+      val runsBefore : Set[Class[_]] = phase.runsBefore.asInstanceOf[Set[Class[_]]]
+
+      orderRequirements.update(phase.getClass, (runsAfter, runsBefore))
+    }
+
+    // propagate ordering constraint : reflexivity
+    pluginPhases.foreach { phase =>
+
+      var (runsAfter, runsBefore) = orderRequirements(phase.getClass)
+
+      // propagate transitive constraints to related phases
+      runsAfter.filter(!isPrimitive(_)).foreach { phaseClass =>
```

Comment:
why do you exclude primitive phases here?

Reply:
Upon reflection, I think we miss a test case, primitive phase should not be a special case.

Reply:
@sjrd : thanks again for the review, I've addressed all raised issues. For this particular one, it turns out the code can be generalized to use the same logic for all already inserted phases. I've updated the code and added a detailed comment:

https://github.com/lampepfl/dotty/pull/3438/files#diff-3226d7606c232d5a280358cb50f2d961R149